Packages/com.chisel.editor/Chisel/Editor/Editor/MaterialBrowser/ChiselMaterialBrowserUtilities.cs

```diff
@@ -0,0 +1,242 @@
+
+            foreach( string id in guids )
+            {
+                ChiselSurfacePresetBrowserTile tile = new ChiselSurfacePresetBrowserTile( id );
+
+                if( labels != null )
+                {
+                    foreach( string label in tile.labels )
+                    {
+                        if(!labels.Contains( label ))
+                            labels.Add( label );
+                    }
+                }
+
+                tiles.Add( tile );
+            }
+        }
+
+        // gets all materials and the labels on them in the project, compares them against a filter,
+        // and then adds them to the list of materials to be used in this window
+        public static void GetMaterials( ref List<ChiselMaterialBrowserTile> materials,
+                                         ref List<ChiselMaterialBrowserTile> usedMaterials,
+                                         ref List<string>                    labels,
+                                         ref List<ChiselModel>               models,
+                                         bool                                usingLabel,
+                                         string                              searchLabel = "",
+                                         string                              searchText  = "" )
+        {
+            if( usingLabel && searchLabel == string.Empty )
+                Debug.LogError( $"usingLabel set to true, but no search term was given. This may give undesired results." );
+
+            materials.ForEach( e =>
+            {
+                e.Dispose();
+                e = null;
+            } );
+
+            materials.Clear();
+
+            // exclude the label search tag if we arent searching for a specific label right now
+            string search = usingLabel ? $"l:{searchLabel} {searchText}" : $"{searchText}";
+
+            string[] guids = AssetDatabase.FindAssets( $"t:Material {search}" );
+
+            // assemble preview tiles
+            foreach( string id in guids )
+            {
+                ChiselMaterialBrowserTile browserTile = new ChiselMaterialBrowserTile( id );
+
+                if( labels != null )
+                {
+                    // add any used labels we arent currently storing
+                    foreach( string label in browserTile.labels )
+                    {
+                        if( !labels.Contains( label ) )
+                            labels.Add( label );
+                    }
+                }
+
+                // check each entry against a filter to exclude certain entries
+                if( IsValidEntry( browserTile ) )
+                {
+                    // if we have the material already, skip, else add it
+                    materials.Add( browserTile );
+                }
+            }
+
+            AssetPreview.SetPreviewTextureCacheSize( materials.Count + 1 );
+
+            models = Object.FindObjectsOfType<ChiselModel>().ToList();
```

Comment:
can also just use AddRange here instead of ToList (after making sure models has been allocated) to prevent some garbage being generated

Reply:
Implemented in next commit.